Fix seedFromFaucet on multiple requests with same lovelace amount

ffakenz opened the pull request:
<!-- Describe your change here -->

## What

When `seedFromFaucet` is called multiple times using the same amount of Lovelace, all subsequent calls yield the same UTxO as the first one repeatedly.

This occurs because `seedFromFaucet` invokes `waitForPayment` on the `receivingAddress` immediately after submitting the transaction for the requested funds, and due to `waitForPayment` selects the UTxO based on the Lovelace amount of value, subsequent calls will repeatedly fetch the first successful submission.

## How

🥐 Now, we wait for the transaction to be submitted on-chain before checking the payment arrived on the receiving address.

🥐 This allows for requesting the faucet for the same amount of Lovelace multiple times.

🥐 Note that the fix only applies when there are previous payments with the same Lovelace amount at the time `seedFromFaucet` is called.

---

<!-- Consider each and tick it off one way or the other -->
* [X] CHANGELOG updated or not needed
* [X] Documentation updated or not needed
* [X] Haddocks updated or not needed
* [X] No new TODOs introduced or explained herafter

ffakenz committed:
Fix seedFromFaucet to wait for tx instead of UTxO on addr

This allows to send same amount multiple times.
Before we were yielding the same value over and over.

ffakenz committed:
Apply fix only if there are old payments with the same lovelace amout

ffakenz committed:
Minor renaming

ffakenz committed:
Reduce timeout pn findOldPaymentsWithSameLovelace

ffakenz committed:
Minor formatting changes

ch1bo commented on hydra-cluster/test/Test/Hydra/Cluster/FaucetSpec.hs:
```diff
@@ -12,44 +12,50 @@ import Hydra.Cluster.Faucet (returnFundsToFaucet, seedFromFaucet, seedFromFaucet
 
 spec :: Spec
 spec = do
-  describe "seedFromFaucet" $
```
Should: keep the `describe` contexts to not need to mention `seedFromFaucet` on each test case

this will also make thee diff here smaller as "seedFromFaucet and returnFundsToFaucet work together" was unchanged.

ch1bo commented on hydra-cluster/test/Test/Hydra/Cluster/FaucetSpec.hs:
```diff
@@ -12,44 +12,50 @@ import Hydra.Cluster.Faucet (returnFundsToFaucet, seedFromFaucet, seedFromFaucet
+            walletVk <- fst <$> generate genKeyPair
+            uTxO1 <- seedFromFaucet node walletVk 2_000_000 (contramap FromFaucet tracer)
+            uTxO2 <- seedFromFaucet node walletVk 2_000_000 (contramap FromFaucet tracer)
+            uTxO1 `shouldNotBe` uTxO2
```
Should: merge this with the upper test

Instead of having two (very similar tests as both request the same amount multiple times), we could just improve the upper one with something like:

```haskell
  utxos <- replicateConcurrently 10 $ do
              vk <- generate genVerificationKey
              seedFromFaucet node vk 1_000_000 (contramap FromFaucet tracer)
  -- 10 unique outputs
  length (fold utxos) `shouldBe` 10
```

ch1bo commented on hydra-cluster/src/CardanoClient.hs:
```diff
@@ -79,23 +79,23 @@ submitTx :: RunningNode -> Tx -> IO ()
   Address ShelleyAddr ->
   IO UTxO
-waitForPayment networkId socket amount addr =
+waitForPayments networkId socket amount addr =
```
Should: document the reason why you renamed this

You seemingly changed the name to indicate this may also return multiple UTxO? You should document that for the future reader.

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Faucet.hs:
```diff
@@ -56,16 +56,38 @@ seedFromFaucet ::
-        submitTransaction networkId nodeSocket (sign faucetSk body)
+        let signedTx = sign faucetSk body
+        submitTransaction networkId nodeSocket signedTx
+        pure signedTx
```
Keep this unchanged

```suggestion
        submitTransaction networkId nodeSocket (sign faucetSk body)
```

ffakenz replied:
but I need the function to return the `signedTx` for calling `awaitTransaction`.

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Faucet.hs:
```diff
@@ -56,16 +56,38 @@ seedFromFaucet ::
+        $ UTxO.pairs newPayments
+    Nothing -> do
+      void $ retryOnExceptions tracer $ submitSeedTx faucetVk faucetSk
+      waitForPayments networkId nodeSocket lovelace receivingAddress
```
Should: This is VERY complicated

Why not just use `awaitTransaction` and then find `theOutput` in it?

```suggestion
   seedTx <- retryOnExceptions tracer $ submitSeedTx faucetVk faucetSk
   producedUTxO <- awaitTransaction networkId nodeSocket seedTx
   pure $ UTxO.filter (== theOutput) producedUTxO
```

ffakenz replied:
worked like a charm!

good catch!

ch1bo reviewed (commented):
I can live with this change, BUT the implementation is so complicated that I had to leave suggestions which should just work as well.

Not having a strong opinion if someone wants to merge this as is though (but please consider my suggestions)

ffakenz committed:
Keep the describe contexts

To not need to mention seedFromFaucet on each test case.

This will also make thee diff smaller.

ffakenz committed:
Add haddock for waitForPayments

ffakenz committed:
Simplify fix logic

ffakenz committed:
Minor formatting change

ffakenz committed:
Merge similar tests into an enhanced one

ffakenz committed:
Minor formatting changes

ch1bo reviewed (approved):
Much better :relieved:
